src/bulk_insert/bulk_insert.c

```diff
@@ -7,388 +7,211 @@
 
 #include "bulk_insert.h"
 #include "../stores/store.h"
+#include <errno.h>
 #include <assert.h>
-
-typedef struct {
-    long long node_count;   // Number of nodes to create under this label.
-    const char *label;      // Label.
-    int label_id;           // ID given to label, recognized by graph object.
-    int attribute_count;    // How may attributes are there for this label.
-    char **attributes;      // Array of attributes.
-} LabelDesc;
-
-void _BulkInsert_Reply_With_Syntax_Error(RedisModuleCtx *ctx, const char* err) {
-    RedisModule_ReplyWithError(ctx, err);
+#include <libgen.h>
+#include "../util/rmalloc.h"
+
+// The first byte of each property in the binary stream
+// is used to indicate the type of the subsequent SIValue
+typedef enum {
+    BI_NULL,
+    BI_BOOL,
+    BI_NUMERIC,
+    BI_STRING
+} TYPE;
+
+// Read the header of a data stream to parse its property keys and update schemas.
+static char** _BulkInsert_ReadHeader(GraphContext *gc, LabelStoreType t,
+                                                  const char *data, size_t *data_idx,
+                                                  int *label_id, unsigned int *prop_count) {
+    // First sequence is entity name 
+    const char *name = data + *data_idx;
```

Comment:
It seems like there's no checks here which verify we do read pass data_len

Reply:
There aren't, we assume that binary token at a minimum contains the label/relation_type name and property count. I'd personally rather make that assumption than risk performance loss from overeager branch execution - see https://github.com/RedisLabsModules/RedisGraph/pull/281#discussion_r241805498